text cursor hover

bytemunch opened the pull request:
closes #46 

An example of how hover-detection and cursor changing can happen. Interaction functions are in userland, should hover join them there or be inbuilt?

PR currently expands interaction fns in userland to add hover bits

bytemunch committed:
text cursor hover example

StaffEngineer commented:
What do you think about having plugin config property to set cursors for `onTyping`, `onHover` and when cursor out of the edit box?

bytemunch commented:
I think there's arguments for and against. For UX it's better to have clear feedback that shows the box is editable, but a lot of games have custom cursors and will want to handle the events on their end. I feel having hover cursors is a sane default, and a config option to provide custom cursors or ignore all cursor modification is a good idea for other use cases.

Not sure about `onTyping` hiding the cursor, do you have a use case for that?

StaffEngineer commented:
@bytemunch type on github text input...giving the comment... it hides the cursor on typing :)

bytemunch commented:
Ah ok, doesn't happen on my Linux, will try with Windows tomorrow and see if I can find out what OSX does. If it's normal UX then emulating it should be default, I agree

StaffEngineer commented:
Oh, I am on mac os, the mouse cursor is hidden on typing _this_ very comment.

bytemunch commented:
Okay, so definitely hide on OSX, Windows is (expectedly) a [mess](https://superuser.com/questions/928839/what-does-the-hide-pointer-while-typing-feature-actually-do) and Linux is... Linux. I think hide on typing by default makes aesthetic sense, can't go wrong following Apple's lead

bytemunch committed:
Merge branch 'main' into hover-cursor

bytemunch committed:
move cursor modify fns to src

caveats:
 - only hides cursor when typing if cursor is in app window
 - only acts on primary window

bytemunch committed:
add config for cursor changes

cursor hide does not work on WASM

bytemunch commented:
added hover events for extensibility with custom cursors, and a config option to disable inbuilt cursor modification

Dimchikkk commented on src/lib.rs:
```diff
@@ -407,7 +419,17 @@ impl Plugin for CosmicEditPlugin {
                 swash_cache: SwashCache::new(),
             })
             .insert_resource(CosmicFontSystem(font_system))
+            // Cursor Bits
```
```suggestion
       
```

Dimchikkk commented on src/lib.rs:
```diff
@@ -407,7 +419,17 @@ impl Plugin for CosmicEditPlugin {
             .add_event::<CosmicTextChanged>();
+
+        // Cursor Bits
+        app.add_systems(Update, (hover_sprites, hover_ui))
```
Let's hide it under the `change_cursor` flag too.

bytemunch replied:
They emit the events for on and off hover, so if user wants to make a custom cursor implementation those fns will still be needed. Maybe more granularity on the options, `change_cursor: default | events | none`?

Dimchikkk replied:
For custom cursors I would prefer plugin config options for cursors.... but `change_cursor` with `default | events | none` also should be fine 👍 I just want to have an option to fully disable it if needed.

bytemunch replied:
Custom cursor configs would be simpler if bevy_cosmic_edit was the only one controlling the cursor, but I feel if there are other plugins or code that also need cursor control hooking an event would be easier to integrate. Full disable a good shout, no need for unused events

Dimchikkk replied:
Makes sense

Dimchikkk commented on src/cursor.rs:
```diff
@@ -0,0 +1,102 @@
+            _ => {}
+        }
+        if let Interaction::Pressed = interaction {
+            commands.insert_resource(Focus(Some(entity)));
```
```suggestion
```
Let's not handle Focus on plugin side.

bytemunch replied:
my bad, copy pasted the fn and missed that bit >.<

Dimchikkk replied:
no prob :)

bytemunch committed:
Update src/cursor.rs

Co-authored-by: StaffEngineer <111751109+StaffEngineer@users.noreply.github.com>

bytemunch committed:
fix focus handling in cursor.rs

bytemunch committed:
Update src/lib.rs

Co-authored-by: StaffEngineer <111751109+StaffEngineer@users.noreply.github.com>

bytemunch committed:
add none | events | default options for cursor

bytemunch committed:
stop sprites sending hover events per frame

bytemunch committed:
commit missed example

Dimchikkk reviewed (approved):
nice!
